lib/core/buffer.hpp

```diff
@@ -41,39 +35,351 @@
+     * and need to pass the metadata down the pipeline.
+     *
+     * It should be called only after acquiring both a full frame to read from
+     * and an empty frame to copy into. Using @c wait_for_full_frame and @c wait_for_empty_frame
+     *
+     * Note it doesn't actually copy the metadata, instead it just copies the pointer
+     * and uses reference counting to track which buffers (and frames) have registered
+     * access to this metadata object.  The stage releases the metadata implicitly
+     * when the @c mark_frame_empty() function is called, which decrements
+     * the reference counter. Once it reaches zero, the the metadata is returned to the
+     * pool.
+     *
+     * @param[in] from_frame_id The frame ID to copy the metadata from
+     * @param[in] to_buf The buffer to copy the metadata into
+     * @param[in] to_frame_id The frame ID in the @c to_buf to copy the metadata into
+     */
+    void pass_metadata(int from_ID, GenericBuffer* to_buf, int to_ID);
+
+    /**
+     * @brief Makes a fully deep copy of the metadata from one object to another
+     *
+     * Unlike pass_metadata this doesn't remove the metadata from the @c from_buf
+     * and requires that the @c to_buf has a metadata object to be copied into.
+     *
+     * @param[in] from_frame_id The frame ID to copy the metadata from
+     * @param[in] to_buf The buffer to copy the metadata into
+     * @param[in] to_frame_id The frame ID in the @c to_buf to copy the metadata into
+     */
+    void copy_metadata(int from_frame_id, GenericBuffer* to_buf, int to_frame_id);
+
+    /// The number of frames kept by this object
+    int num_frames;
+
+    /// Should we shut down (stop returning new frames)?
+    bool shutdown_signal;
+
+    /// The list of consumer names registered to this buffer
+    std::map<std::string, StageInfo> consumers;
+
+    /// The list of producer names registered to this buffer
+    std::map<std::string, StageInfo> producers;
+
+    /// The name of the buffer for use in debug messages.
+    std::string buffer_name;
+
+    /// The type of the buffer for use in writing data.
+    std::string buffer_type;
+
+    /// The pool of info objects
+    metadataPool* metadata_pool;
+
+    /// Array of buffer info objects, for tracking information about each buffer.
+    std::vector<metadataContainer*> metadata;
+
+protected:
+    /// The main lock for frame state management
+    std::recursive_mutex mutex;
+
+    /// The condition variable for calls to @c wait_for_full_buffer
+    std::condition_variable_any full_cond;
+
+    /// The condition variable for calls to @c wait_for_empty_buffer
+    std::condition_variable_any empty_cond;
+
+    void private_copy_metadata(int dest_frame_id, GenericBuffer* src, int src_frame_id);
+};
+
+/**
+ * @brief A buffer to manage the signalling between stages when those
+ * stages want to communicate data using a ring buffer.
+ *
+ * This class is a little strange in that it doesn't actually allocate
```

Comment:
Perhaps we should name it something like `SignalRingBuffer`?  Since I could imagine having a ring buffer in CPU space which actually manages the underlying memory.

Reply:
Good point -- I think I will hold off on renaming until a later PR, after I've added the multi-consumer, since other things may shake out in that process.